File: java/150.evaluate-reverse-polish-notation.java
/*
 * @lc app=leetcode id=150 lang=java
 *
 * [150] Evaluate Reverse Polish Notation
 */
class Solution {
    public int evalRPN(String[] tokens) {
        if (tokens == null) return 0;
        Stack<Integer> stack = new Stack<>();
        Map<String, Operator> operateMap = new HashMap<>();
        operateMap.put("+", new Plus());
        operateMap.put("-", new Minus());
        operateMap.put("*", new Times());
        operateMap.put("/", new Divide());
        for (String token : tokens) {
            if (operateMap.containsKey(token)) {
                int val2 = stack.pop(), val1 = stack.pop();
                stack.push(operateMap.get(token).operate(val1, val2));
            } else {
                stack.push(Integer.parseInt(token));
            }
        }
        return stack.pop();
    }
}

interface Operator {
    int operate(int val1, int val2);
}

class Plus implements Operator {
    public int operate(int val1, int val2) {
        return val1 + val2;
    }
}

class Minus implements Operator {
    public int operate(int val1, int val2) {
        return val1 - val2;
    }
}

class Times implements Operator {
    public int operate(int val1, int val2) {
        return val1 * val2;
    }
}

class Divide implements Operator {
    public int operate(int val1, int val2) {
        return val1 / val2;
    }
}

File: go/150.evaluate-reverse-polish-notation.go
/*
 * @lc app=leetcode id=150 lang=golang
 *
 * [150] Evaluate Reverse Polish Notation
 */

// @lc code=start
func evalRPN(tokens []string) int {
    if tokens == nil || len(tokens) == 0 {
		return 0
	}
	sk := make([]int, len(tokens)/2+1)
	skp := -1
	for _, tk := range tokens {
		switch tk {
		case "+", "-", "*", "/":
			val := op(tk, sk[skp-1], sk[skp])
			skp = skp-1
			sk[skp] = val
		default:
			skp++
			sk[skp], _ = strconv.Atoi(tk)
		}
	}
	return sk[0]
}

func op(tk string, v1, v2 int) (val int) {
	switch tk {
	case "+":
		val = v1 + v2
	case "-":
		val = v1 - v2
	case "*":
		val = v1 * v2
	case "/":
		val = v1 / v2
	}
	return
}
// @lc code=end

